README.md
[![Circle CI](https://circleci.com/gh/Griesbacher/check_prometheus/tree/master.svg?style=svg)](https://circleci.com/gh/Griesbacher/check_prometheus/tree/master)
[![Go Report Card](https://goreportcard.com/badge/github.com/Griesbacher/check_prometheus)](https://goreportcard.com/report/github.com/Griesbacher/check_prometheus)
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](http://www.gnu.org/licenses/gpl-3.0)

# check_prometheus
Monitoring Plugin to check the health of a Prometheus server and its data

## Usage
### Global Options
```
$ check_prometheus -h
NAME:
   check_prometheus - Checks different prometheus stats as well the data itself
   Copyright (c) 2017 Philip Griesbacher
   https://github.com/Griesbacher/check_prometheus

USAGE:
   check_prometheus [global options] command [command options] [arguments...]

VERSION:
   0.0.1

COMMANDS:
     mode, m  check mode
     help, h  Shows a list of commands or help for one command

GLOBAL OPTIONS:
   -t value       Seconds till check returns unknown, 0 to disable (default: 10)
   -f value       If the checked data is older then this in seconds, unknown will be returned. Set to 0 to disable. (default: 300)
   --help, -h     show help
   --version, -v  print the version
```

### Command options

```
$ check_prometheus mode -h
   NAME:
      check_prometheus mode - check mode
   
   USAGE:
      check_prometheus mode command [command options] [arguments...]
   
   COMMANDS:
        ping, p         Returns the build informations
        query, q        Checks collected data
        targets_health  Returns the health of the targets
   
   OPTIONS:
      --help, -h  show help
```

### Subcommand options example

```
$ check_prometheus mode query -h
NAME:
   check_prometheus mode query - Checks collected data

USAGE:
   check_prometheus mode query [command options] [arguments...]

DESCRIPTION:
   Your Promqlquery has to return a vector / scalar / matrix result. The warning and critical values are applied to every value.
   Examples:
       Vector:
           check_prometheus mode query -q 'up'
       --> OK - Query: 'up'|'up{instance="192.168.99.101:9245", job="iapetos"}'=1;;;; 'up{instance="0.0.0.0:9091", job="prometheus"}'=1;;;;

       Scalar:
           check_prometheus mode query -q 'scalar(up{job="prometheus"})'
       --> OK - OK - Query: 'scalar(up{job="prometheus"})' returned: '1'|'scalar'=1;;;;

       Matrix:
           check_prometheus mode query -q 'http_requests_total{job="prometheus"}[5m]'
       --> OK - Query: 'http_requests_total{job="prometheus"}[5m]'

       Search and Replace:
           check_prometheus m query -q 'up' --search '.*job=\"(.*?)\".*' --replace '$1'
       --> OK - Query: 'up'|'prometheus'=1;;;; 'iapetos'=0;;;;

           check_prometheus m q -q '{handler="prometheus",quantile="0.99",job="prometheus",__name__=~"http_.*bytes"}' --search '.*__name__=\"(.*?)\".*' --replace '$1' -a 'http_in_out'
       --> OK - Alias: 'http_in_out'|'http_request_size_bytes'=296;;;; 'http_response_size_bytes'=5554;;;;

OPTIONS:
   --address value  Prometheus address: Protocol + IP + Port. (default: "http://localhost:9100")
   -q value         Query to be executed
   -a value         Alias, will replace the query within the output, if set
   -w value         Warning value. Use nagios-plugin syntax here.
   -c value         Critical value. Use nagios-plugin syntax here.
   --search value   If this variable is set, the given Golang regex will be used to search and replace the result with the 'replace' flag content. This will be appied on the perflabels.
   --replace value  See search flag. If the 'search' flag is empty this flag will be ignored.

```

Changelog.md
# 0.0.1 - 27.09.2016
## Features:
- Ping mode
- Query mode
- Targets_Health mode

mode/ping.go
package mode

import (
	"context"
	"encoding/json"
	"fmt"
	"github.com/griesbacher/check_prometheus/helper"
	"github.com/griesbacher/check_x"
	"github.com/prometheus/common/model"
	"time"
)

type buildInfo struct {
	Metric struct {
		Name      string `json:"__name__"`
		Branch    string `json:"branch"`
		Goversion string `json:"goversion"`
		Instance  string `json:"instance"`
		Job       string `json:"job"`
		Revision  string `json:"revision"`
		Version   string `json:"version"`
	} `json:"metric"`
	Value []interface{} `json:"value"`
}

// Ping will fetch build information from the prometheus server
func Ping(address string) (err error) {
	apiClient, err := helper.NewAPIClientV1(address)
	if err != nil {
		return
	}
	query := `prometheus_build_info{job="prometheus"}`
	startTime := time.Now()
	result, err := apiClient.Query(context.TODO(), query, time.Now())
	endTime := time.Now()
	if err != nil {
		return
	}
	vector := result.(model.Vector)
	if len(vector) != 1 {
		return fmt.Errorf("The query '%s' did not return a vector with a single entry", query)
	}
	sample := vector[0]
	helper.CheckTimestampFreshness(sample.Timestamp)
	jsonBytes, err := sample.MarshalJSON()
	if err != nil {
		return
	}
	var dat buildInfo
	if err = json.Unmarshal(jsonBytes, &dat); err != nil {
		return
	}
	check_x.NewPerformanceData("duration", endTime.Sub(startTime).Seconds()).Unit("s").Min(0)
	check_x.Exit(check_x.OK, fmt.Sprintf("Version: %s, Instance %s", dat.Metric.Version, dat.Metric.Instance))
	return err
}


mode/query.go
package mode

import (
	"context"
	"fmt"
	"github.com/griesbacher/check_prometheus/helper"
	"github.com/griesbacher/check_x"
	"github.com/prometheus/common/model"
	"regexp"
	"strconv"
	"time"
)

//Query allows the user to test data in the prometheus server
func Query(address, query, warning, critical, alias, search, replace string) (err error) {
	warn, err := check_x.NewThreshold(warning)
	if err != nil {
		return
	}

	crit, err := check_x.NewThreshold(critical)
	if err != nil {
		return
	}
	var re *regexp.Regexp
	if search != "" {
		re, err = regexp.Compile(search)
		if err != nil {
			return
		}
	}

	apiClient, err := helper.NewAPIClientV1(address)
	if err != nil {
		return
	}

	result, err := apiClient.Query(context.TODO(), query, time.Now())
	if err != nil {
		return
	}
	switch result.Type() {
	case model.ValScalar:
		scalar := result.(*model.Scalar)
		scalarValue := float64(scalar.Value)
		helper.CheckTimestampFreshness(scalar.Timestamp)

		check_x.NewPerformanceData(replaceLabel("scalar", re, replace), scalarValue).Warn(warn).Crit(crit)
		state := check_x.Evaluator{Warning: warn, Critical: warn}.Evaluate(scalarValue)

		resultAsString := strconv.FormatFloat(scalarValue, 'f', -1, 64)
		if alias == "" {
			check_x.Exit(state, fmt.Sprintf("Query: '%s' returned: '%s'", query, resultAsString))
		} else {
			check_x.Exit(state, fmt.Sprintf("Alias: '%s' returned: '%s'", alias, resultAsString))
		}
	case model.ValVector:
		vector := result.(model.Vector)
		states := check_x.States{}
		for _, sample := range vector {
			helper.CheckTimestampFreshness(sample.Timestamp)

			sampleValue := float64(sample.Value)
			check_x.NewPerformanceData(replaceLabel(model.LabelSet(sample.Metric).String(), re, replace), sampleValue).Warn(warn).Crit(crit)
			states = append(states, check_x.Evaluator{Warning: warn, Critical: warn}.Evaluate(sampleValue))

		}
		return evalStates(states, alias, query)
	case model.ValMatrix:
		matrix := result.(model.Matrix)
		states := check_x.States{}
		for _, sampleStream := range matrix {
			for _, value := range sampleStream.Values {
				helper.CheckTimestampFreshness(value.Timestamp)
				states = append(states, check_x.Evaluator{Warning: warn, Critical: warn}.Evaluate(float64(value.Value)))
			}
		}
		return evalStates(states, alias, query)
	default:
		check_x.Exit(check_x.Unknown, fmt.Sprintf("The query did not return a suppoted type(scalar, vector, matrix), instead: '%s'. Query: '%s'", result.Type().String(), query))
		return nil
	}
	return err
}

func replaceLabel(label string, re *regexp.Regexp, replace string) string {
	if re != nil {
		label = re.ReplaceAllString(label, replace)
	}
	return label
}

func evalStates(states check_x.States, alias, query string) error {
	state, err := states.GetWorst()
	if err != nil {
		return err
	}
	if alias == "" {
		check_x.Exit(*state, fmt.Sprintf("Query: '%s'", query))
	} else {
		check_x.Exit(*state, fmt.Sprintf("Alias: '%s'", alias))
	}
	return nil
}


helper/prometheus.go
package helper

import (
	"fmt"
	"github.com/griesbacher/check_x"
	"github.com/prometheus/client_golang/api"
	"github.com/prometheus/client_golang/api/prometheus/v1"
	"github.com/prometheus/common/model"
	"io/ioutil"
	"net/http"
	"time"
)

//NewAPIClientV1 will create an prometheus api client v1
func NewAPIClientV1(address string) (v1.API, error) {
	client, err := api.NewClient(api.Config{
		Address: address,
	})
	if err != nil {
		return nil, err
	}
	return v1.NewAPI(client), nil
}

//DoAPIRequest does the http handling for an api request
func DoAPIRequest(address string) ([]byte, error) {
	resp, err := http.DefaultClient.Get(address)
	if err != nil {
		return nil, err
	}
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	return body, nil
}

//TimestampFreshness is the amount of second a result is treated as valid
var TimestampFreshness int

//CheckTimestampFreshness tests if the data is still valid
func CheckTimestampFreshness(timestamp model.Time) {
	CheckTimeFreshness(time.Unix(int64(timestamp), 0))
}

//CheckTimeFreshness tests if the data is still valid
func CheckTimeFreshness(timestamp time.Time) {
	if TimestampFreshness == 0 {
		return
	}
	timeDiff := time.Now().Sub(timestamp)
	if int(timeDiff.Seconds()) > TimestampFreshness {
		check_x.Exit(check_x.Unknown, fmt.Sprintf("One of the scraped data exceed the freshness by %ds", int(timeDiff.Seconds())-TimestampFreshness))
	}
}


mode/targetsHealth.go
package mode

import (
	"encoding/json"
	"fmt"
	"github.com/griesbacher/check_prometheus/helper"
	"github.com/griesbacher/check_x"
	"net/url"
	"path"
	"time"
)

const (
	//DefaultLabel is used if the given label is wrong
	DefaultLabel = "instance"
)

type targets struct {
	Status string `json:"status"`
	Data   struct {
		ActiveTargets []struct {
			DiscoveredLabels struct {
				Address     string `json:"__address__"`
				MetricsPath string `json:"__metrics_path__"`
				Scheme      string `json:"__scheme__"`
				Job         string `json:"job"`
			} `json:"discoveredLabels"`
			Labels     map[string]string `json:"labels"`
			ScrapeURL  string            `json:"scrapeUrl"`
			LastError  string            `json:"lastError"`
			LastScrape time.Time         `json:"lastScrape"`
			Health     string            `json:"health"`
		} `json:"activeTargets"`
	} `json:"data"`
}

func getTargets(address string) (*targets, error) {
	u, err := url.Parse(address)
	if err != nil {
		return nil, err
	}
	u.Path = path.Join(u.Path, "/api/v1/targets")
	jsonBytes, err := helper.DoAPIRequest(u.String())
	if err != nil {
		return nil, err
	}
	var dat targets
	if err = json.Unmarshal(jsonBytes, &dat); err != nil {
		return nil, err
	}
	return &dat, nil
}

//TargetsHealth tests the health of the targets
func TargetsHealth(address, label, warning, critical string) (err error) {
	warn, err := check_x.NewThreshold(warning)
	if err != nil {
		return
	}

	crit, err := check_x.NewThreshold(critical)
	if err != nil {
		return
	}

	targets, err := getTargets(address)
	if err != nil {
		return
	}
	if (*targets).Status != "success" {
		return fmt.Errorf("The API target returnstatus was %s", (*targets).Status)
	}
	msg := ""
	healthy := 0
	unhealthy := 0
	for _, target := range (*targets).Data.ActiveTargets {
		msg += fmt.Sprintf("Job: %s, Instance: %s, Health: %s, Last Error: %s\n", target.Labels["job"], target.Labels["instance"], target.Health, target.LastError)
		health := 0.0
		if target.Health != "up" {
			health = 1
			unhealthy += 1
		} else {
			healthy += 1
		}
		if val, ok := target.Labels[label]; ok {
			check_x.NewPerformanceData(val, health)
		} else {
			check_x.NewPerformanceData(target.Labels[DefaultLabel], health)
		}
	}
	var healthRate float64
	sumTargets := float64(len((*targets).Data.ActiveTargets))
	if sumTargets == 0 {
		healthRate = 0
	} else {
		healthRate = float64(healthy) / sumTargets
	}
	check_x.NewPerformanceData("health_rate", healthRate).Warn(warn).Crit(crit).Min(0).Max(1)
	check_x.NewPerformanceData("targets", sumTargets).Min(0)
	state := check_x.Evaluator{Warning: warn, Critical: warn}.Evaluate(healthRate)
	check_x.LongExit(state, fmt.Sprintf("There are %d healthy and %d unhealthy targets", healthy, unhealthy), msg)
	return
}


main.go
package main

import (
	"github.com/griesbacher/check_prometheus/helper"
	"github.com/griesbacher/check_prometheus/mode"
	"github.com/griesbacher/check_x"
	"github.com/urfave/cli"
	"os"
	"time"
)

var (
	address  string
	timeout  int
	warning  string
	critical string
	query    string
	alias    string
	search   string
	replace  string
	label    string
)

func startTimeout() {
	if timeout != 0 {
		check_x.StartTimeout(time.Duration(timeout) * time.Second)
	}
}

func main() {
	app := cli.NewApp()
	app.Name = "check_prometheus"
	app.Usage = "Checks different prometheus stats as well the data itself\n   Copyright (c) 2017 Philip Griesbacher\n   https://github.com/Griesbacher/check_prometheus"
	app.Version = "0.0.1"
	flagAddress := cli.StringFlag{
		Name:        "address",
		Usage:       "Prometheus address: Protocol + IP + Port.",
		Destination: &address,
		Value:       "http://localhost:9100",
	}
	flagWarning := cli.StringFlag{
		Name:        "w",
		Usage:       "Warning value. Use nagios-plugin syntax here.",
		Destination: &warning,
	}
	flagCritical := cli.StringFlag{
		Name:        "c",
		Usage:       "Critical value. Use nagios-plugin syntax here.",
		Destination: &critical,
	}
	flagQuery := cli.StringFlag{
		Name:        "q",
		Usage:       "Query to be executed",
		Destination: &query,
	}
	flagAlias := cli.StringFlag{
		Name:        "a",
		Usage:       "Alias, will replace the query within the output, if set",
		Destination: &alias,
	}
	flagLabel := cli.StringFlag{
		Name:        "l",
		Usage:       "Prometheus-Label, which will be used for the performance data label. By default job and instance should be available.",
		Destination: &label,
		Value:       mode.DefaultLabel,
	}
	app.Commands = []cli.Command{
		{
			Name:    "mode",
			Aliases: []string{"m"},
			Usage:   "check mode",
			Subcommands: []cli.Command{
				{
					Name:        "ping",
					Aliases:     []string{"p"},
					Usage:       "Returns the build informations",
					Description: `This check requires that the prometheus server itself is listetd as target. Following query will be used: 'prometheus_build_info{job="prometheus"}'`,
					Action: func(c *cli.Context) error {
						startTimeout()
						return mode.Ping(address)
					},
					Flags: []cli.Flag{
						flagAddress,
					},
				}, {
					Name:    "query",
					Aliases: []string{"q"},
					Usage:   "Checks collected data",
					Description: `Your Promqlquery has to return a vector / scalar / matrix result. The warning and critical values are applied to every value.
   Examples:
       Vector:
           check_prometheus mode query -q 'up'
       --> OK - Query: 'up'|'up{instance="192.168.99.101:9245", job="iapetos"}'=1;;;; 'up{instance="0.0.0.0:9091", job="prometheus"}'=1;;;;

       Scalar:
           check_prometheus mode query -q 'scalar(up{job="prometheus"})'
       --> OK - OK - Query: 'scalar(up{job="prometheus"})' returned: '1'|'scalar'=1;;;;

       Matrix:
           check_prometheus mode query -q 'http_requests_total{job="prometheus"}[5m]'
       --> OK - Query: 'http_requests_total{job="prometheus"}[5m]'

       Search and Replace:
           check_prometheus m query -q 'up' --search '.*job=\"(.*?)\".*' --replace '$1'
       --> OK - Query: 'up'|'prometheus'=1;;;; 'iapetos'=0;;;;

           check_prometheus m q -q '{handler="prometheus",quantile="0.99",job="prometheus",__name__=~"http_.*bytes"}' --search '.*__name__=\"(.*?)\".*' --replace '$1' -a 'http_in_out'
       --> OK - Alias: 'http_in_out'|'http_request_size_bytes'=296;;;; 'http_response_size_bytes'=5554;;;;`,

					Action: func(c *cli.Context) error {
						startTimeout()
						return mode.Query(address, query, warning, critical, alias, search, replace)
					},
					Flags: []cli.Flag{
						flagAddress,
						flagQuery,
						flagAlias,
						flagWarning,
						flagCritical,
						cli.StringFlag{
							Name:        "search",
							Usage:       "If this variable is set, the given Golang regex will be used to search and replace the result with the 'replace' flag content. This will be appied on the perflabels.",
							Destination: &search,
						},
						cli.StringFlag{
							Name:        "replace",
							Usage:       "See search flag. If the 'search' flag is empty this flag will be ignored.",
							Destination: &replace,
						},
					},
				}, {
					Name:        "targets_health",
					Usage:       "Returns the health of the targets",
					Description: `The warning and critical thresholds are appied on the health_rate. The health_rate is calculted: sum(healthy) / sum(targets).`,
					Action: func(c *cli.Context) error {
						startTimeout()
						return mode.TargetsHealth(address, label, warning, critical)
					},
					Flags: []cli.Flag{
						flagAddress,
						flagWarning,
						flagCritical,
						flagLabel,
					},
				},
			},
		},
	}
	app.Flags = []cli.Flag{
		cli.IntFlag{
			Name:        "t",
			Usage:       "Seconds till check returns unknown, 0 to disable",
			Value:       10,
			Destination: &timeout,
		},
		cli.IntFlag{
			Name:        "f",
			Usage:       "If the checked data is older then this in seconds, unknown will be returned. Set to 0 to disable.",
			Value:       300,
			Destination: &helper.TimestampFreshness,
		},
	}

	if err := app.Run(os.Args); err != nil {
		check_x.ErrorExit(err)
	}
}
